src/ripple/core/tests/Coroutine.test.cpp

```diff
@@ -0,0 +1,109 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/test/jtx.h>
+#include <ripple/core/JobCoro.h>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace ripple {
+namespace test {
+
+class Coroutine_test : public beast::unit_test::suite
+{
+public:
+    void
+    test_coroutine()
+    {
+        using namespace jtx;
+        Env env(*this);
+        std::atomic<int> i{0};
+        std::condition_variable cv;
+        auto& jq = env.app().getJobQueue();
+        jq.setThreadCount(0, true);
+        jq.postCoro(jtCLIENT, "Coroutine-Test",
+            [&](std::shared_ptr<JobCoro> jc)
+            {
+                std::thread t(
+                    [&i, jc]()
+                    {
+                        std::this_thread::sleep_for(
+                            std::chrono::milliseconds(10));
+                        ++i;
+                        jc->resume();
+                    });
+                jc->yield();
+                t.join();
+                ++i;
+                cv.notify_one();
+            });
+
+        {
+            std::mutex m;
+            std::unique_lock<std::mutex> lk(m);
+            cv.wait_for(lk, std::chrono::milliseconds(1000),
+                [&]() {return i == 2;});
+        }
+        jq.shutdown();
+        expect(i == 2);
+    }
+
+    void
+    test_incorrect_order()
+    {
+        using namespace jtx;
+        Env env(*this);
+        std::atomic<int> i{0};
+        std::condition_variable cv;
+        auto& jq = env.app().getJobQueue();
+        jq.setThreadCount(0, true);
+        jq.postCoro(jtCLIENT, "Coroutine-Test",
+            [&](std::shared_ptr<JobCoro> jc)
+            {
+                jc->resume();
+                jc->yield();
```

Comment:
A coroutine can suspend and then resume execution, that is all it does. The order is simply reversed here, resume is called before yield, which is okay but effectively does not yield. I am not sure a comment to explain that is necessary in this unit test, but certainly as @vinniefalco mentioned I need to add better documentation to the Coroutine class.